lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,169 @@
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+from coursewarehistoryextended.fields import UnsignedBigIntAutoField
+from xmodule_django.models import BlockTypeKeyField, CourseKeyField
+
+import json
+
+
+class VisibleBlocksModel(models.Model):
+    """
+    A django model used to track the state of a set of visibile blocks under a given subsection at the time they are
+    used for grade calculation.
+
+    This state is represented using an array of serialized BlockRecords, stored in the blocks_json field. A
+    hash of this json array is used for lookup purposes.
+
+    Note that while blocks_json may be accessed directly (this is python, we're all consenting adults here), the
+    blocks property is provided to make deserialization more convenient. Similarly, hashed should not be set directly.
```

Comment:
Is this something we can roughly enforce via quality checks? I'm guessing we probably have something like "if variable has leading underscore, it's private, and shouldn't be accessed except for internally".